### PR TITLE
Fix data linking to allow contour over-plotting

### LINKED ISSUE #318: Can't plot contours from Moment Map Plugin over cube viewer
Here is the workflow

1. Read in continuum-subtracted MANGA data cube
2. Select subset in spectrum viewer
3. Create Moment 0 map in Moment Maps Plugin
4. Select Flux Cube and  Moment0 Data Set in Cube viewer
5. Select Moment0 Layer
6. Activate Contours

Contours for the same Moment0 product can be viewed in a separate image viewer with no problem,
but they don't appear in the cube viewer. The same contour levels are generated in the cube viewer, 
but they aren't visible:

manga_moment1_contours

### PR DESCRIPTION
**Pull request by javerbukh**
### Description

This pull request is to address the problem with over-plotting contours in cubeviz. It uses `pixel_component_ids` to do this, so data is linked by pixel rather than by WCS.

Fixes #318 

Close #1159

GIF shows contours of cubes on top of each other.
Mar-09-2022 09-35-03

GIF shows contours of moment map plotted over a cube.
Mar-09-2022 09-36-13

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [x] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Did the CI pass? If not, are the failures related?
- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [x] Is a milestone set? Milestone is only currently required for PRs related to Imviz MVP.
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

### CODE REVIEW

File: jdaviz/app.py [hunk 1]
```diff
@@ -311,16 +311,39 @@ def _link_new_data(self):
 
         # Can't link if there's no world_component_ids
         wc_new = self.data_collection[-1].world_component_ids
-        if wc_new == []:
-            return
+        pc_new = self.data_collection[-1].pixel_component_ids
+
+        # if wc_new == []:
+        #     return
```
Comment (javerbukh) [inline jdaviz/app.py:317]: I'm not sure what the consequences are of commenting out this line are. I need to do it so that moment map data (which does not have `world_component_ids`) can be linked to the rest of the data in `data_collection`.

File: jdaviz/configs/cubeviz/plugins/moment_maps/moment_maps.py [hunk 2]
```diff
@@ -126,7 +126,8 @@ def vue_calculate_moment(self, *args):
         label = "Moment {}: {}".format(n_moment, self._selected_data.label)
         fname_label = self._selected_data.label.replace("[", "_").replace("]", "")
         self.filename = "moment{}_{}.fits".format(n_moment, fname_label)
-        self.data_collection[label] = self.moment
```
Comment (javerbukh) [inline jdaviz/configs/cubeviz/plugins/moment_maps/moment_maps.py:129]: I changed this so it would be consistent with the rest of the plugins, let me know if it needs to be reverted for whatever reason.

**Comment by javerbukh** [comment]
Using `get_object()` to determine if data is from the moment map plugin does not play well with Imviz - unless I use a try/except. Is there another way to determine if data comes from that plugin besides parsing the label?

**Comment by pllim** [comment]
Maybe you can move the linking code into the plugin? Does it have to be at app level?

**Comment by pllim** [comment]
👎 to Imviz going the `except` route. People already complaining enough about performance as it is.

**Comment by javerbukh** [comment]
@pllim It could be plugin level rather than app, but I think I prefer to keep as much linking in one place as possible. I could definitely be convinced otherwise though.

File: jdaviz/app.py [hunk 3]
```diff
@@ -311,16 +311,43 @@ def _link_new_data(self):
...
         for i in range(new_len - 1):
             wc_old = self.data_collection[i].world_component_ids
+            pc_old = self.data_collection[i].pixel_component_ids
+
             if wc_old == []:
                 continue
             else:
-                self.data_collection.add_link(LinkSame(wc_old[0], wc_new[0]))
+                # New data is a moment map
+                if "Moment" in self.data_collection[-1].label:
```
[the remark this answers is not in the document]
Comment (javerbukh) [reply jdaviz/app.py]: Good call! That is the method I am using in the latest commit.

**Comment by javerbukh** [comment]
Does this fix #1399 as well?

**Comment by pllim** [comment]
Issue 1399 in which repo? We are not that high yet here.

**Comment by javerbukh** [comment]
Sorry, I saw the "JDAT-1399" in the title of that issue and ignored the "JDAT" part 🙈

**Review by duytnguyendtn — APPROVED** [review]
Once I actually understood what I was doing, contours worked as according to what I was expecting! Thanks for the improvement; I finally feel like I don't have to walk on eggshells  here!

**Comment by pllim** [comment]
@javerbukh , I added NDData support (both 2D and 3D) to #1040 . Do you think that would help your cause here?

**Comment by javerbukh** [comment]
@pllim I don't think that applies directly to this PR. I think there would have to be a follow-up to your PR that converts moment map data into NDData, which may be its own can of worms that I do not want to unpack here.

File: jdaviz/app.py [hunk 4]
```diff
@@ -311,16 +311,31 @@ def _link_new_data(self):
 
         # Can't link if there's no world_component_ids
         wc_new = self.data_collection[-1].world_component_ids
-        if wc_new == []:
```
Comment (pllim) [inline jdaviz/app.py:314]: This was added in #435 by @rosteen to improve performance. Can we be sure that performance will not degrade by removing this line?

File: jdaviz/app.py [hunk 5]
```diff
@@ -311,16 +311,31 @@ def _link_new_data(self):
...
         for i in range(new_len - 1):
             wc_old = self.data_collection[i].world_component_ids
+            pc_old = self.data_collection[i].pixel_component_ids
+
             if wc_old == []:
                 continue
             else:
-                self.data_collection.add_link(LinkSame(wc_old[0], wc_new[0]))
+                # Link cubes to each other
+                if len(pc_old) > 2 and len(pc_new) > 2:
```
Comment (pllim) [inline jdaviz/app.py:325]: Do we actually still need this now that you are linking in the plugin?

**Comment by pllim** [comment]
Re: #530 -- I do not think this fixes that issue. Linking was not the problem (and I already forgot what exactly PO wanted that I couldn't make possible). Just overlaying contour from another image on top image is already possible in Imviz in `main`, see https://github.com/glue-viz/glue-jupyter/issues/231#issuecomment-1064358692

File: CHANGES.rst [hunk 6]
```diff
@@ -39,6 +39,8 @@ Bug Fixes
 Cubeviz
 ^^^^^^^
 
+- Fixed linking of data to allow contour over-plotting. [#1154]
```
Comment (pllim) [inline CHANGES.rst:42]:
```suggestion
- Fixed linking of data to allow contour over-plotting for moment map. [#1154]
```

File: jdaviz/app.py (hunk 4 above)
Comment (javerbukh) [reply jdaviz/app.py:314]: The problem is that for moment map data, `world_component_ids` is empty, since it seems like it does not receive spatial WCS infomation ([see related comment](https://github.com/spacetelescope/jdaviz/blob/main/jdaviz/configs/cubeviz/plugins/moment_maps/tests/test_moment_maps.py#L31)). Otherwise I could keep that check in.

File: jdaviz/app.py (hunk 5 above)
Comment (javerbukh) [reply jdaviz/app.py:325]: Yes, in main cubes also cannot utilize contour over-plotting because they are not linked by pixels in the x and y coordinates.

File: jdaviz/configs/cubeviz/plugins/moment_maps/moment_maps.py [hunk 7]
```diff
@@ -146,3 +155,27 @@ def vue_save_as_fits(self, *args):
...
         self.hub.broadcast(SnackbarMessage(
             f"Moment map saved to {os.path.abspath(self.filename)}", sender=self, color="success"))
+
+    def _link_moment_data(self):
+        new_len = len(self.app.data_collection)
+
+        # Can't link if there's no world_component_ids
+        pc_new = self.app.data_collection[-1].pixel_component_ids
+
+        # Link to the first dataset with compatible coordinates
```
[the remark this answers is not in the document]
Comment (javerbukh) [reply jdaviz/configs/cubeviz/plugins/moment_maps/moment_maps.py:165]: Not sure what you mean here. For contouring to work (at least to my understanding, @astrofrog can correct me if I'm wrong), the data must be linked by pixels.

[the remark this answers is not in the document]
Comment (pllim) [reply jdaviz/configs/cubeviz/plugins/moment_maps/moment_maps.py:165]: Really? Doesn't WCS linking translate to pixel space anyway at some point?

File: CHANGES.rst (hunk 6 above)
Comment (javerbukh) [reply CHANGES.rst:42]: It's not just for moment maps though, it does not work in main using cubes. For example, contours of VAR over FLUX in a single viewer is not possible in main, but it is in this PR.

File: jdaviz/app.py (hunk 5 above)
Comment (pllim) [reply jdaviz/app.py:325]: Now I am super confused... Haha. What is the one in the plugin for then?

File: jdaviz/app.py (hunk 4 above)
Comment (pllim) [reply jdaviz/app.py:314]: Is it more correct, then, to pass that WCS info to moment output instead of all this code?

**Comment by javerbukh** [comment]
@pllim I'm going to answer the questions in one place:

> Is it more correct, then, to pass that WCS info to moment output instead of all this code?

I think that relates to your other question about WCS linking translating to pixel space. I do not think that is the case, in cubeviz at least. While passing WCS info into moment map data is good, I think it relates more to the "Improve moment map plugin" (i.e., the same issue of converting moment map to use `NDData`) issue than to this one.

> What is the one in the plugin for then?

Because the devs agreed that the code specifically for linking moment map data to other jdaviz data should exist in the moment map plugin and not in `app.py`, so there is some code duplication.

> Really? Doesn't WCS linking translate to pixel space anyway at some point?

I think that's how it works in imviz but I do not think it works like that in the rest of jdaviz.


I think I should add that this PR not only allows moment map contour over-plotting, but ALL contour over-plotting. I think somewhere over the last year we made changes that stopped that functionality from working as expected, and this PR returns that expected functionality.

File: CHANGES.rst (hunk 6 above)
Comment (pllim) [reply CHANGES.rst:42]: oh ok

File: jdaviz/app.py (hunk 5 above)
Comment (pllim) [reply jdaviz/app.py:325]: Okay, I think I get it... this one here is for cube on cube... ?

**Comment by javerbukh** [comment]
That's weird, I don't know why anything merged recently would impact this PR. The instructions for contouring are [here](https://jdaviz.readthedocs.io/en/latest/cubeviz/plugins.html#contours). As for your PR, it only works in the case of superimposing moment map contours on the data it is based off of. I think there would need to be a PR that fixes app-wide linking before that PR is merged. I guess I can make a new issue for fixing linking and then we can split that from the moment map work, but everything is already fixed in this PR...

File: jdaviz/app.py (hunk 5 above)
Comment (javerbukh) [reply jdaviz/app.py:325]: Correct!

File: jdaviz/configs/default/plugins/gaussian_smooth/tests/test_gaussian_smooth.py [hunk 8]
```diff
@@ -20,10 +20,10 @@ def test_linking_after_spectral_smooth(spectrum1d_cube):
 
     assert len(dc) == 2
     assert dc[1].label == 'Smoothed test stddev 3.2'
-    assert len(dc.external_links) == 1
+    assert len(dc.external_links) == 3
```
Comment (pllim) [inline jdaviz/configs/default/plugins/gaussian_smooth/tests/test_gaussian_smooth.py:23]: Why is this changed? This PR does not touch any Gaussian Smooth stuff, does it?

**Review by pllim — APPROVED** [review]
Okay, not sure what happened last night but the cube-on-cube overlay works now (even after rebase). Imviz example notebook seems to work and Imviz contour overlay still works, so I guess this is good to go in.

I'll leave this open for a bit more in case other devs want to review, especially if they are worries about Mosviz performance or something, given the deletion of the code added to fix performance.

File: jdaviz/configs/default/plugins/gaussian_smooth/tests/test_gaussian_smooth.py (hunk 8 above)
Comment (javerbukh) [reply jdaviz/configs/default/plugins/gaussian_smooth/tests/test_gaussian_smooth.py:23]: This PR adds more links (pixel x and y) for cubes in addition to linking for moment map, so that is where the extra 2 `external_links` come from.

**Review by kecnry — APPROVED** [review]
I'll trust the previous code reviews as I'm not too familiar with linking-logic.  I tested in cubeviz and couldn't find any way to break it!  Even works during slicing when the "bitmap" is hidden and only contours shown.

**Comment by pllim** [comment]
If a dev cannot find a way to break things, it is a very good sign. Let's merge so everyone can have a good weekend. Thanks! 😸
